src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -929,125 +725,316 @@ internal ThreadLocalList(int ownerThreadId)
             {
                 _ownerThreadId = ownerThreadId;
             }
+
+            /// <summary>Gets whether the list is empty.</summary>
+            internal bool IsEmpty => _headIndex >= _tailIndex;
+
             /// <summary>
             /// Add new item to head of the list
             /// </summary>
             /// <param name="item">The item to add.</param>
             /// <param name="updateCount">Whether to update the count.</param>
-            internal void Add(T item, bool updateCount)
+            internal void LocalPush(T item, bool needsSync)
             {
-                checked
-                {
-                    _count++;
-                }
-                Node node = new Node(item);
-                if (_head == null)
-                {
-                    Debug.Assert(_tail == null);
-                    _head = node;
-                    _tail = node;
-                    _version++; // changing from empty state to non empty state
-                }
-                else
+                Debug.Assert(Environment.CurrentManagedThreadId == _ownerThreadId);
+                bool lockTaken = false;
+                try
                 {
-                    node._next = _head;
-                    _head._prev = node;
-                    _head = node;
+                    // Full fence to ensure subsequent reads don't get reordered before this
+                    Interlocked.Exchange(ref _currentOp, (int)ConcurrentBagListOperation.Add);
```

Comment:
I believe the concern is that, with just the acquire/release semantics from the volatile reads/writes, it's theoretically possible for the subsequent reads of _tailIndex and _headIndex to move up before the write to _currentOp.  If that were to happen, a steal could end up mucking with _headIndex without waiting for this in-flight local push to quiesce, even if _headIndex and _tailIndex were close enough together that the push and steal should have synchronized with each other.  I'll add a comment.